FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1323 @@
+          Ref StackValue2 = LoadStackValueAtOffset_Slow(offset);
+          if (ReducedPrecisionMode) {
+            CmpNode = IREmit->_FCmp(8, StackValue1, StackValue2);
+          } else {
+            CmpNode = IREmit->_F80Cmp(StackValue1, StackValue2);
+          }
+        } else {
+          if (ReducedPrecisionMode) {
+            CmpNode = IREmit->_FCmp(8, StackMember1.StackDataNode, StackMember2.StackDataNode);
+          } else {
+            CmpNode = IREmit->_F80Cmp(StackMember1.StackDataNode, StackMember2.StackDataNode);
+          }
+        }
+        IREmit->ReplaceUsesWithAfter(CodeNode, CmpNode, CodeNode);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+      case OP_F80STACKTEST: {
+        const auto* Op = IROp->C<IROp_F80StackTest>();
+        auto offset = Op->SrcStack;
+        const auto& [Valid, StackMember] = StackData.top(offset);
+
+        MigrateToSlowPathIf(Valid != StackSlot::VALID);
+
+        Ref CmpNode {};
+        Ref ZeroConst = IREmit->_VCastFromGPR(ReducedPrecisionMode ? 8 : 16, 8, GetConstant(0));
+
+        if (SlowPath) {
+          auto* StackValue = LoadStackValueAtOffset_Slow(offset);
+          if (ReducedPrecisionMode) {
+            CmpNode = IREmit->_FCmp(8, StackValue, ZeroConst);
+          } else {
+            CmpNode = IREmit->_F80Cmp(StackValue, ZeroConst);
+          }
+        } else {
+          if (ReducedPrecisionMode) {
+            CmpNode = IREmit->_FCmp(8, StackMember.StackDataNode, ZeroConst);
+          } else {
+            CmpNode = IREmit->_F80Cmp(StackMember.StackDataNode, ZeroConst);
+          }
+        }
+        IREmit->ReplaceUsesWithAfter(CodeNode, CmpNode, CodeNode);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+
+      case OP_F80CMPVALUE: {
+        const auto* Op = IROp->C<IROp_F80CmpValue>();
+        const auto& Value = CurrentIR.GetNode(Op->X80Src);
+        const auto& [Valid, StackMember] = StackData.top();
+
+        MigrateToSlowPathIf(Valid != StackSlot::VALID);
+
+        Ref CmpNode {};
+        if (SlowPath) {
+          auto* StackValue = LoadStackValueAtTop_Slow();
+          if (ReducedPrecisionMode) {
+            CmpNode = IREmit->_FCmp(8, StackValue, Value);
+          } else {
+            CmpNode = IREmit->_F80Cmp(StackValue, Value);
+          }
+        } else {
+          if (ReducedPrecisionMode) {
+            CmpNode = IREmit->_FCmp(8, StackMember.StackDataNode, Value);
+          } else {
+            CmpNode = IREmit->_F80Cmp(StackMember.StackDataNode, Value);
+          }
+        }
+
+        IREmit->ReplaceUsesWithAfter(CodeNode, CmpNode, CodeNode);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_F80XTRACTSTACK: {
+        const auto& [Valid, StackMember] = StackData.top();
+
+        MigrateToSlowPathIf(Valid != StackSlot::VALID);
+
+        if (SlowPath) {
+          Ref st0 = LoadStackValueAtTop_Slow();
+          Ref Exp {};
```

Comment:
Probably want to `st0 = SlowPath ? Load() : StackDataNode` outside the if, then you can deduplicate the splitting logic and just have the if for writing it out